core/finality-grandpa/src/tests.rs

```diff
@@ -114,6 +113,17 @@ impl TestNetFactory for GrandpaTestNet {
 		(shared_import.clone(), Some(shared_import), Mutex::new(Some(link)))
 	}
 
+	fn add_peer(&mut self, config: &ProtocolConfig) {
```

Comment:
Code duplication. If this is the same implementation, can we not just move this as default implementation into the trait definition?

Reply:
Since I added `NetworkSpecialization` as a type on the `TestNetFactory`, we cannot have a default `add_peer` method on the factory because it contains `let specialization = DummySpecialization;`, which requires an implementation of the TesnetFactory trait to specificy `type Specialization = DummySpecialization;`. 

I did remove as much duplication as possible by adding the `create_peer` function, which is used in each implementation.

Reply:
Introduce a trait 
```
trait Specialization {
    fn create() -> Self;
}
```

And then `Self::Specialization::create()`.

Reply:
I'm not convinced an additional trait is better than some duplication. I also expect most testnet implementations will have their own setup for peer creation and this duplication will prove a non-issue.

Reply:
You already repeated the same code 4 times.

Reply:
Duplication is just duplicated work for any change.

Reply:
It's very easy to copy paste  code, it was very hard for me to actually add `NetworkSpecialization` to `TestNetFactory` because `add_peer` was implemented on `TestNetFactory`, as opposed to duplicated for each testnet implementation. I'd rather avoid this kind of complexity going forward.

Reply:
It is always easy to copy paste code ^^ But copy-pasting is very error-prone. 

I don't really understand the rest of your comment. Sorry^^ 

So before this was already a default implementation and the only thing that prevents us from doing the same again, is the initialization of the specialization?

Reply:
I added a `fn make_specialization(&self) -> Self::Specialization` method to `TestNetFactory` and then moved the duplicate code from `add_peer` to the `TestNetFactory` trait. Do you want me to push these changes?

Reply:
Sounds better than what I proposed! Nice @andresilva :+1:

Reply:
Ok, you're right the proposed trait is a better solution.

Reply:
You can remove this implementation now :)

Reply:
@andresilva just seeing your comment, feel free to push your version if it is better than what I just committed...

Reply:
Nope, yours is fine :+1: